clang/test/SemaOpenCL/invalid-kernel-attrs.cl

```diff
@@ -42,8 +42,8 @@ kernel __attribute__((intel_reqd_sub_group_size(-1))) void kernel16() {} // expe
 kernel __attribute__((intel_reqd_sub_group_size(8))) __attribute__((intel_reqd_sub_group_size(16))) void kernel17() {} //expected-warning{{attribute 'intel_reqd_sub_group_size' is already applied with different parameters}}
 
 __kernel __attribute__((work_group_size_hint(8,-16,32))) void neg1() {} //expected-error{{'work_group_size_hint' attribute requires a non-negative integral compile time constant expression}}
-__kernel __attribute__((reqd_work_group_size(8,16,-32))) void neg2(){} // expected-error{{'reqd_work_group_size' attribute requires a non-negative integral compile time constant expression}}
+__kernel __attribute__((reqd_work_group_size(8, 16, -32))) void neg2() {} // OK
```

Comment:
Is it intentional that attribute now can accept negative arguments?

Reply:
Yes, it is intentional that the attribute accepts negative values.

The old behavior was that the attribute accepted unsigned integer values, so a negative number would be treated as an unsigned value anyway. New code behaves the same -- accept any integer value, and when getting the dimension values (for error checking or for codegen), get them as unsigned values.

We do not diagnose negative values (because those would just be treated as positive values because the argument is unsigned) here with template support, but we would still diagnose things that aren't an integer constant expression (like 1.2f).

Reply:
> The old behavior was that the attribute accepted unsigned integer values, so a negative number would be treated as an unsigned value anyway. 

I don't fully understand this statement. It seems that negative numbers were diagnosed with an error. Am I missing something?
I'm a bit worried that we don't diagnose negative values anymore. I believe it was done on purpose, since even though negative values are converted to unsigned ones, `reqd_work_group_size` attribute roughly means required (sorry everyone who knows OpenCL) number of threads to execute particular kernel, and negative value doesn't make sense here.

Reply:
Thanks for the explanation. I will fix that. just to confirm here: 
__kernel __attribute__((reqd_work_group_size(8, 16, 4294967294))) void ok1() {} --> this was OK, 
if we add diagnostic for negative arguments, this will give warning now? is it expected?

Reply:
The comment in the code below seems explaining this. 4294967294 will be negative for signed values due to overflow, but the maximum positive value for unsigned values is bigger than for signed values, so such big number should be still ok here, since we expect unsigned value as an argument for the attribute.

Reply:
> I don't fully understand this statement. It seems that negative numbers were diagnosed with an error. Am I missing something? I'm a bit worried that we don't diagnose negative values anymore. I believe it was done on purpose, since even though negative values are converted to unsigned ones, reqd_work_group_size attribute roughly means required (sorry everyone who knows OpenCL) number of threads to execute particular kernel, and negative value doesn't make sense here.

I think there's some confusion here. The old definition in Attr.td said that the attribute takes three unsigned arguments. The old semantic checks in SemaDeclAttr.cpp used to then bizarrely diagnose negative values. There did not appear to be any reason for this odd behavior -- because the arguments are unsigned, there is not supposed to be the notion of a negative argument value in the first place.

Now that we're updating the code to allow expressions, we're trying to ensure that we get the correct behavior despite this confusion. I agree that negative values do not make sense as attribute arguments here -- but that tells me the attribute argument should continue to be unsigned values. It might make sense to warn on sign conversion of an attribute argument, but I can't understand why it should be a hard error. That's like causing an error on `unsigned i = -1;`.

Reply:
Put a slightly different way -- if there's a range of valid values that this can take as arguments, we should diagnose when a value is out of range (and using an error is fine for this). e.g., if a max attribute says the values are 1, 5, 10 and the reqd attribute uses the values 1, 4, -1, the diagnostic should be because the value 0xFFFFFFFF'FFFFFFFF is greater than 10, not because -1 is less than 0. However, if a max attribute says the values are 1, 5, and 0xFFFFFFFF'FFFFFFFF, there's no reason to reject if the reqd values are 1, 4, -1, but maybe it may still make sense to warn the user that they've got a signed/unsigned mismatch on the third argument.

Reply:
I think warning about sign conversion would make sense here.

Reply:
def warn_attribute_requires_positive_integer : Warning<
  "%0 attribute requires a %select{positive|non-negative}1 "
  "integral compile time constant expression - attribute ignored">,
 InGroup<IgnoredAttributes>;

def err_attribute_requires_positive_integer : Error<
  "%0 attribute requires a %select{positive|non-negative}1 "
  "integral compile time constant expression">;

Currently we are giving warning for negative attribute value and ignore the attribute. Just to confirm with you @Fznamznon and @AaronBallman whether we should ignore the attribute.

Reply:
Done

Reply:
GitHub told me about a question on this thread but for whatever reason, the GitHub UI refuses to show it to me. :-/

> Currently we are giving warning for negative attribute value and ignore the attribute. Just to confirm with you @Fznamznon and @AaronBallman whether we should ignore the attribute.

Warning on sign mismatches seems fine to me, but we should not ignore the attribute in that case (just keep treating the value as being unsigned).

Reply:
Thanks @AaronBallman for the confirmation. That means we need to add new one and do not ignore the attribute.

def warn_attribute_requires_positive_integer : Warning<
  "%0 attribute requires a %select{positive|non-negative}1 "
  "integral compile time constant expression - attribute ignored">,
 InGroup<IgnoredAttributes>;

Reply:
I have added new diagnostic with new group since it doe not match with anyone. Also added test to show that the accributes accept  negative argument value with a warning.

Reply:
> @Fznamznon, are you OK with it if we do not ignore the attribute here for negative values with a warning?

I'm ok with that.

Reply:
Thanks @Fznamznon for the confirmation.